### PR TITLE
Feature/memcache

### PR DESCRIPTION
**Pull request by ArneGudermann**
This PR implements the memcache for viur-datastore. 
TThe idea behind this is to make queries that often run over db.Get more performant. For example, with each request the session is loaded which can now be cached in memcache. This makes the loading up to **7-15** times faster

Todos:

- [x] Make a PR for the core to add `appengine-python-standard` to  the dependencies and wirte a wrapper for the main app.
- [x] Make a PR for the viur-base for the app.yml
- [ ] Make  a PR for the app_server to support the local development.

### CODE REVIEW

**Comment by ArneGudermann** [comment]
the problem is that for memcache there is only a local development environment for python 2. maybe we can discuss this in viur round

File: src/viur/datastore/__init__.py [hunk 1]
```diff
@@ -70,4 +71,5 @@
 	"UnavailableError",
 	"NoMutationResultsError",
 	"is_viur_datastore_request_ok",
+	"cache"
```
Comment (sveneberth) [inline src/viur/datastore/__init__.py:74]:
```suggestion
	"cache",
```

File: src/viur/datastore/cache.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import sys
+from typing import Any, Dict, List, Union
+
+from viur.datastore.config import conf
+from viur.datastore.types import Entity, Key
+
+memcache_max_batch_size = 30
+memcache_namespace = "viur-datastore"
+memcache_timeout = 60 * 60
+memcache_max_size = 1_000_000
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:10]: They look like constants. So I would write them in UPPER_SNAKE_CASE.

```suggestion
MEMCACHE_MAX_BATCH_SIZE = 30
MEMCACHE_NAMESPACE = "viur-datastore"
MEMCACHE_TIMEOUT = 60 * 60
MEMCACHE_MAX_SIZE = 1_000_000
```

File: src/viur/datastore/cache.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def get(keys: Union[str, Key, List[str], List[Key]]):
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:13]:
```suggestion
def get(keys: Union[str, Key, List[str], List[Key]]) -> dict:
```

File: src/viur/datastore/cache.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	return res
+
+
+def set(cache_data: Union[Dict, List[Entity]]):
+	if isinstance(cache_data, dict):
+		# Add only values to cache <= memcache_max_size (1.000.000)
+		cache_data = {str(key): value for key, value in cache_data.items() if get_size(value) <= memcache_max_size}
+	elif isinstance(cache_data, list):
+		# Add only values to cache <= memcache_max_size (1.000.000)
+		cache_data = {str(item.key): item for item in cache_data if get_size(item) <= memcache_max_size}
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:31]: less duplicate code and adds a type check. (not tested :grin:)
```suggestion
    if isinstance(cache_data, list):
        cache_data = {item.key: item for item in cache_data}
    elif not isinstance(cache_data, dict):
        raise TypeError(f"Invalid type {type(cache_data)}. Expected a list or dict.")
    # Add only values to cache <= memcache_max_size (1.000.000)
    cache_data = {str(key): value for key, value in cache_data.items()
                  if get_size(value) <= memcache_max_size}
```

File: src/viur/datastore/cache.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	keys = [str(key) for key in keys]  # Enforce that all keys are strings
+	res = {}
+
+	while keys:
+		res |= conf["memcache_client"].get_multi(keys[:memcache_max_batch_size], namespace=memcache_namespace)
+		keys = keys[memcache_max_batch_size:]
+	return res
+
+
+def set(cache_data: Union[Dict, List[Entity]]):
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:25]:
```suggestion
def set(cache_data: Union[Dict[Key, Entity], List[Entity]]) -> None:
```

File: src/viur/datastore/cache.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		# Add only values to cache <= memcache_max_size (1.000.000)
+		cache_data = {str(item.key): item for item in cache_data if get_size(item) <= memcache_max_size}
+	keys = list(cache_data.keys())
+	while keys:
+		data = {key: cache_data[key] for key in keys[:memcache_max_batch_size]}
+		conf["memcache_client"].set_multi(data, namespace=memcache_namespace, time=memcache_timeout)
+		keys = keys[memcache_max_batch_size:]
+
+
+def delete(keys: Union[str, Key, List[str], List[Key]]):
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:39]:
```suggestion
def delete(keys: Union[str, Key, List[str], List[Key]]) -> None:
```

File: src/viur/datastore/cache.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def get_size(obj: Any) -> int:
+	"""
+		Utility function that counts the size of an object in bytes.
+	"""
+	if isinstance(obj, dict):
+		return sum([get_size([k, v]) for k, v in obj.items()])
+	elif isinstance(obj, list):
+		return sum([get_size(x) for x in obj])
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:55]: generator comprehension should be a little bit faster than a list comprehension.

```suggestion
		return sum(get_size([k, v]) for k, v in obj.items())
	elif isinstance(obj, list):
		return sum(get_size(x) for x in obj)
```

File: src/viur/datastore/transport.pyx [hunk 8]
```diff
@@ -586,26 +584,42 @@ def Get(keys: Union[Key, List[Key]]) -> Union[None, Entity, List[Entity]]:
...
+	res_from_db = {}
+	untouched_keys = copy(keys)
+	while keys:
+		keys_for_request = keys[:300]
+
+		if conf["use_memcache_client"]:
+			res_from_cache = cache.get(keys_for_request)
+			# Convert the keys back to "class" representation
+			res_from_cache = {Key.from_legacy_urlsafe(key): value for key, value in
+							  res_from_cache.items()}
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:605]: I think this is more readable:
```suggestion
			res_from_cache = {Key.from_legacy_urlsafe(key): value 
			                  for key, value in res_from_cache.items()}
```

File: src/viur/datastore/transport.pyx [hunk 9]
```diff
@@ -615,15 +629,27 @@ def Get(keys: Union[Key, List[Key]]) -> Union[None, Entity, List[Entity]]:
...
+			res_from_db.update(toEntityStructure(element.at_key("found"), isInitial=True))
+
 		if (element.at_pointer("/deferred").error() == SUCCESS):
-			keyList = toPythonStructure(element.at_key("deferred")) + keyList[300:]
+			deferred_keys = toPythonStructure(element.at_key("deferred"))
+			missing_keys_paths=[keyToPath(key) for key in missing_keys]
+			keys = keys[300:]
+			for i,deferred_key in enumerate(deferred_keys):
+				if deferred_key in missing_keys_paths:
+					keys.insert(0,missing_keys[i])
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:640]: The `viur-datastore` needs a automatically linter ... spaces are no crime

```suggestion
			deferred_keys = toPythonStructure(element.at_key("deferred"))
			missing_keys_paths = [keyToPath(key) for key in missing_keys]
			keys = keys[300:]
			for i, deferred_key in enumerate(deferred_keys):
				if deferred_key in missing_keys_paths:
					keys.insert(0, missing_keys[i])
```

File: src/viur/datastore/transport.pyx [hunk 10]
```diff
@@ -677,6 +703,8 @@ def Delete(keys: Union[Key, List[Key], Entity, List[Entity]]) -> None:
 		if arrayElem.size() != abs(len(keys)):
 			print(req.content)
 			raise ValueError("Invalid number of mutation-results received")
+	if conf["use_memcache_client"] :
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:706]:
```suggestion
	if conf["use_memcache_client"]:
```

**Review by sveneberth — CHANGES_REQUESTED** [review]

File: src/viur/datastore/cache.py (hunk 4 above)
Comment (ArneGudermann) [reply src/viur/datastore/cache.py:31]: Mhh i think i must rework  this part

File: src/viur/datastore/cache.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+	This Module controls the Interaction with the Memcache from Google
+	To activate the cache copy this code in your main.py
+	..  code-block:: python
+	# Example
+
+	if not conf["viur.instance.is_dev_server"]:
+		from google.appengine.api.memcache import Client
+		from viur.core import db
+		db.config["use_memcache_client"] = True
```
Comment (sveneberth) [inline src/viur/datastore/cache.py:24]:
```suggestion
```

File: src/viur/datastore/transport.pyx [hunk 12]
```diff
@@ -736,6 +764,9 @@ def Put(entities: Union[Entity, List[Entity]]) -> Union[Entity, List[Entity]]:
 			entities[idx].version = toPyStr(innerArrayElem.at_key("version").get_string())
 			preincrement(arrayIt)
 			idx += 1
+		if conf["use_memcache_client"]:
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:767]:
```suggestion
		if conf["memcache_client"] is not None:
```

File: src/viur/datastore/transport.pyx [hunk 13]
```diff
@@ -677,6 +703,8 @@ def Delete(keys: Union[Key, List[Key], Entity, List[Entity]]) -> None:
 		if arrayElem.size() != abs(len(keys)):
 			print(req.content)
 			raise ValueError("Invalid number of mutation-results received")
+	if conf["use_memcache_client"]:
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:706]:
```suggestion
	if conf["memcache_client"] is not None:
```

File: src/viur/datastore/transport.pyx [hunk 14]
```diff
@@ -615,15 +629,27 @@ def Get(keys: Union[Key, List[Key]]) -> Union[None, Entity, List[Entity]]:
...
+			keys = keys[300:]
+			for i, deferred_key in enumerate(deferred_keys):
+				if deferred_key in missing_keys_paths:
+					keys.insert(0, missing_keys[i])
+
 		else:
-			keyList = keyList[300:]
+			keys = keys[300:]
+	res = res_from_db | res_from_cache
+	if conf["use_memcache_client"]:
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:645]:
```suggestion
	if conf["memcache_client"] is not None:
```

File: src/viur/datastore/transport.pyx [hunk 15]
```diff
@@ -586,26 +584,42 @@ def Get(keys: Union[Key, List[Key]]) -> Union[None, Entity, List[Entity]]:
...
 	res = {}
-	while keyList:
-		requestedKeys = keyList[:300]
+	res_from_cache = {}
+	res_from_db = {}
+	untouched_keys = copy(keys)
+	while keys:
+		keys_for_request = keys[:300]
+
+		if conf["use_memcache_client"]:
```
Comment (sveneberth) [inline src/viur/datastore/transport.pyx:601]:
```suggestion
		if conf["memcache_client"] is not None:
```

File: src/viur/datastore/config.py [hunk 16]
```diff
@@ -22,5 +22,9 @@
 		"RESOURCE_EXHAUSTED",
 		"UNAUTHENTICATED",
 		"UNAVAILABLE"
-	}
+	},
+	# Set this to 'True' and the viur data store will cache the reads and writes.
+	"use_memcache_client": False,
```
Comment (sveneberth) [inline src/viur/datastore/config.py:27]:
```suggestion
```

**Review by sveneberth — CHANGES_REQUESTED** [review]
I think the option `use_memcache_client` is superfluous. It is enough to set the client. If you set it, you obviously want to use the memcache, otherwise not.

**Review by skoegl — APPROVED** [review]

**Pull request merged**